lib/trino-filesystem-cache-alluxio/src/main/java/io/trino/filesystem/alluxio/AlluxioInputHelper.java

```diff
@@ -0,0 +1,174 @@
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+package io.trino.filesystem.alluxio;
+
+import alluxio.client.file.CacheContext;
+import alluxio.client.file.URIStatus;
+import alluxio.client.file.cache.CacheManager;
+import alluxio.client.file.cache.PageId;
+import alluxio.conf.AlluxioConfiguration;
+import alluxio.conf.PropertyKey;
+import com.google.common.primitives.Ints;
+import io.trino.filesystem.Location;
+
+import java.io.IOException;
+import java.nio.ByteBuffer;
+
+import static com.google.common.base.Preconditions.checkArgument;
+import static java.lang.Integer.max;
+import static java.lang.Math.min;
+import static java.util.Objects.requireNonNull;
+
+// Inspired by https://github.com/Alluxio/alluxio/blob/4e39eda0305a0042edaeae649b503b4508623619/dora/core/client/fs/src/main/java/alluxio/client/file/cache/LocalCacheFileInStream.java#L50
+// We implement a variant of this class to enable positioned reads
+public class AlluxioInputHelper
+{
+    private final URIStatus status;
+    private final CacheManager cacheManager;
+    private final CacheStats statistics;
+    private final Location location;
+    private final int pageSize;
+    private final long fileLength;
+    private final int bufferSize;
+    private final byte[] readBuffer;
+
+    private long bufferStart;
+    private long bufferEnd;
+
+    public AlluxioInputHelper(Location location, URIStatus status, CacheManager cacheManager, AlluxioConfiguration configuration, CacheStats statistics)
+    {
+        this.status = requireNonNull(status, "status is null");
+        this.fileLength = status.getLength();
+        this.cacheManager = requireNonNull(cacheManager, "cacheManager is null");
+        this.pageSize = (int) requireNonNull(configuration, "configuration is null").getBytes(PropertyKey.USER_CLIENT_CACHE_PAGE_SIZE);
+        this.statistics = requireNonNull(statistics, "statistics is null");
+        this.location = requireNonNull(location, "location is null");
+        // Buffer to reduce the cost of doing page aligned reads
+        this.bufferSize = pageSize;
+        this.readBuffer = new byte[bufferSize];
+    }
+
+    public int doBufferRead(long position, byte[] bytes, int offset, int length)
+    {
+        if (length == 0) {
+            return 0;
+        }
+        if (position < bufferStart || position >= bufferEnd) {
+            return 0;
+        }
+        int bytesToCopy = min(length, Ints.saturatedCast(Long.max(bufferEnd - position, 0)));
+        System.arraycopy(readBuffer, Ints.saturatedCast(position - bufferStart), bytes, offset, bytesToCopy);
+        return bytesToCopy;
+    }
+
+    public int doCacheRead(long position, byte[] bytes, int offset, int length)
+            throws IOException
+    {
+        int remainingLength = length;
+        while (remainingLength > 0) {
+            int bytesReadFromCache = readPageFromCache(position, bytes, offset, remainingLength);
```

Comment:
Let's say that we wanted to read 10MB from cache and the cache has the first and last 2MB but not the middle 6MB.
Will we bail here after reading the first 2MB and then read the rest 8MB from the external source or we will we get 4MB from cache and 6MB from external source ?

Reply:
We will bail after reading the first 2MB ad read the remaining 8MB from the external source.